toonz/sources/image/ffmpeg/tiio_gif.cpp

```diff
@@ -0,0 +1,246 @@
+
+#include "tsystem.h"
+#include "tiio_gif.h"
+#include "trasterimage.h"
+#include "timageinfo.h"
+#include <QStringList>
+
+//===========================================================
+//
+//  TImageWriterGif
+//
+//===========================================================
+
+class TImageWriterGif : public TImageWriter {
+public:
+  int m_frameIndex;
+
+  TImageWriterGif(const TFilePath &path, int frameIndex, TLevelWriterGif *lwg)
+      : TImageWriter(path), m_frameIndex(frameIndex), m_lwg(lwg) {
+    m_lwg->addRef();
+  }
+  ~TImageWriterGif() { m_lwg->release(); }
+
+  bool is64bitOutputSupported() override { return false; }
+  void save(const TImageP &img) override { m_lwg->save(img, m_frameIndex); }
+
+private:
+  TLevelWriterGif *m_lwg;
+};
+
+//===========================================================
+//
+//  TLevelWriterGif;
+//
+//===========================================================
+
+TLevelWriterGif::TLevelWriterGif(const TFilePath &path, TPropertyGroup *winfo)
+    : TLevelWriter(path, winfo) {
+  if (!m_properties) m_properties = new Tiio::GifWriterProperties();
+  std::string scale = m_properties->getProperty("Scale")
+                          ->getValueAsString();
+  m_scale = QString::fromStdString(scale).toInt();
+  TBoolProperty *looping =
+      (TBoolProperty *)m_properties->getProperty("Looping");
+  m_looping = looping->getValue();
+  TBoolProperty *palette =
+      (TBoolProperty *)m_properties->getProperty("Generate Palette");
+  m_palette    = palette->getValue();
+  ffmpegWriter = new Ffmpeg();
+  ffmpegWriter->setPath(m_path);
+  // m_frameCount = 0;
+  if (TSystem::doesExistFileOrLevel(m_path)) TSystem::deleteFile(m_path);
+}
+
+//-----------------------------------------------------------
+
+TLevelWriterGif::~TLevelWriterGif() {
+  QStringList preIArgs;
+  QStringList postIArgs;
+  QStringList palettePreIArgs;
+  QStringList palettePostIArgs;
+
+  int outLx = m_lx;
+  int outLy = m_ly;
+
+  // set scaling
+  outLx = m_lx * m_scale / 100;
+  outLy = m_ly * m_scale / 100;
+  //ffmpeg doesn't like resolutions that aren't divisible by 2.
+  if (outLx % 2 != 0) outLx++;
+  if (outLy % 2 != 0) outLy++;
+
+  QString palette;
+  QString filters = "fps=" + QString::number(m_frameRate) + ",scale=" +
+                    QString::number(outLx) + ":-1:flags=lanczos";
```

Comment:
I think you should remove the "fps" filter here and use "-r" input option instead.

To be more precise, please modify the lines 74-75 to be as follows;
`QString filters = "scale=" + QString::number(outLx) + ":-1:flags=lanczos";`
...and insert following lines to the line 93;
```
preIArgs << "-r";
preIArgs << QString::number(m_frameRate);
```


When I checked your PR, I found some problem: If you set a scene fps to other than 25, some frames are dropped in the output GIF. For example, when you render a scene with 20 frames, the result GIF becomes 19 frames under 24fps (= the default value of OpenToonz) and 11 frames under 12.5fps.
According to [the documentation of ffmpeg filters](https://ffmpeg.org/ffmpeg-filters.html#fps-1), it says;

> ## fps
>Convert the video to specified constant frame rate **by duplicating or dropping frames** as necessary.

I'm new to ffmpeg so I'm not 100% convinced, but this option may be the cause of this problem.
The "-r" option is documented [here](https://ffmpeg.org/ffmpeg.html#Video-Options) as follows:

> ### -r[:stream_specifier] _fps (input/output,per-stream)_
> Set frame rate (Hz value, fraction or abbreviation).
> 
> As an input option, ignore any timestamps stored in the file and instead generate timestamps assuming constant frame rate fps.

Here is a comparison of the rendered results. Both results are rendered from the same scene which has 20 frames in length and 8 fps frame rate. Please note that the some frames are dropped from the former result. 

- With "fps" filter (current implementation):
![using_fps_filter_8fps](https://cloud.githubusercontent.com/assets/17974955/17171200/8d70c846-542a-11e6-8c44-1718a47ddf28.gif)
 
- With "-r" input option (my suggestion):
![using_-r_option_8fps](https://cloud.githubusercontent.com/assets/17974955/17171237/bc15226e-542a-11e6-9e51-0a7e0d64b1a5.gif)